rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,760 @@
+
+##### `tsh piv agent`
+
+`tsh piv agent` will be made available as a hidden command, primarily for
+development. If in the future we get requests to fully support this command,
+we may make it a public command and make any necessary improvements.
+
+For now, we will not put exorbitant effort into providing a good UX with this command
+(e.g. stealing focus for hardware key prompts, re-login on cert expiration),
+but it should be fully functional.
+
+##### Agent already running
+
+If there is already a hardware key agent running at `$TEMP/.Teleport-PIV/agent.sock`,
+a newly started agent client would fail to open a new listener on that same
+socket. Instead of failing, the new client will ping the running agent to check
+if it's active.
+
+If the ping request fails, the socket will be treated as abandoned and
+automatically replaced by the new agent.
+
+If the ping request succeeds, an error will be returned:
+
+```console
+> tsh piv agent
+Error: another agent instance is already running. PID: 86946.
+```
+
+Note: this should be a very uncommon edge case, as it would only occur if the
+user already has `tsh piv agent` running or another instance of Teleport connect.
+
+Note: with Teleport Connect, this error would be displayed when it attempts
+to start the agent, but Teleport Connect would not fail to start. The error
+would be shown in Teleport Connect's debug logs.
+
+##### Running the agent before login
+
+The hardware key agent does not depend on Teleport login state, meaning a user
+can run it before logging in.
+
+For the login itself, Teleport clients will interface directly with the
+hardware key to check/generate the private key. Then, the client will interface
+with the key through the hardware key agent that is already running.
+
+#### Dependent clients
+
+Dependent clients should interact with the hardware key agent seamlessly so
+that there is limited UX degradation when compared to the a client connecting
+directly to the hardware key.
+
+##### Agent stops running
+
+If the agent stops unexpectedly during a dependent client's operation, it may
+lead to an error for the dependent client. Rather than returning this error,
+the client will log the error as debug and try again by connecting directly to
+the hardware key.
+
+Note: Teleport Connect and `tsh piv agent` will both attempt to restart the
+agent if it runs into an error, with backoff.
+
+#### Hardware key prompts
+
+The agent is [responsible for prompting hardware key PIN and touch](#pin-and-touch-prompts)
+on behalf of dependent clients.
+
+Note: sometimes, Teleport connect uses `tsh` commands directly, which prompts
+for PIN/touch via the CLI within a Teleport Connect console (e.g. `tsh ssh`).
+A side effect of this change is that the teleport key agent will now prompt via
+Teleport Connect on behalf of these `tsh` commands within Teleport Connect.
+This is ultimately a benefit as the previous CLI prompts will be replaced with
+the UI prompts of Teleport Connect.
```

Comment:
@ravicious I didn't think of this until I noticed it while testing Teleport Connect. WDYT about this side effect? Is it a good thing, or is there a reason we weren't using the UI prompts for connecting to a server?